sql/item_strfunc.cc

```diff
@@ -1348,13 +1348,31 @@ bool Item_func_sformat::fix_length_and_dec()
 
   for (uint i=0 ; i < arg_count ; i++)
   {
-    char_length+= args[i]->max_char_length();
+    // scan the number in {:d} and add it to the char_length
+    size_t num= 0;
+    const char* str= args[i]->name.str;
+    char delimiter1[]= "{:";
+    char delimiter2[]= "}";
+
+    const char* pos1 = str;
+    const char* pos2 = NULL;
+
+    char num_str[100]; // allocate a buffer for the number string
+    while ((pos1 = strstr(pos1, delimiter1)) != NULL &&
+           (pos2 = strstr(pos1 + strlen(delimiter1), delimiter2)) != NULL) {
+        strncpy(num_str, pos1 + strlen(delimiter1), pos2 - pos1 - strlen(delimiter1));
+        num_str[pos2 - pos1 - strlen(delimiter1)] = '\0';
+        num+= strtol(num_str, NULL, 10);
```

Comment:
`strtol` with a not-`NULL` end bufferptr would avoid the need to use a `num_str` buffer.